### ISSUE TITLE
Memory leaks with EmguCV.Stitching.Stitcher.Stitch()

### ISSUE THREAD

**Issue by DrSilbermann** [emgucv/emgucv#256]
I use the Emgu.CV.Stiching Class for the stitching algorithm, an although i try to dispose the resources correctly i get huge memory leaks, depending on the amount and size of the images i use. 

I am using VS 2015 and EmguCV Version 4.1.0.3420 and running a Debug Build.
Here is a minimal reproducible example of my code:

    using System;
    using System.IO;

    using Emgu.CV;
    using Emgu.CV.Structure;
    using Emgu.CV.Stitching;
    using Emgu.CV.Util;
    using Emgu.CV.Features2D;
    using Emgu.CV.CvEnum;

    namespace Example
    {
        class Program
        {
            static void Main(string[] args)
            {
                string folderPath = @"enter your folder with images here";       //Paste your the path of your folder, which contains the images, here       
                string format = ".jpg";
                string[] filePaths = Directory.GetFiles(folderPath, ("*" + format));
                string storagePath;

                Directory.CreateDirectory(folderPath + @"\Stitched");                                               //creates new Directionary if not exist
                int num = 1;                                                                                        //Sets file numbering
                while (true)                                                                                        //Checks if there is already an exisitng file with that name and change numbering "num" if "true"
                {
                    storagePath = folderPath + @"\Stitched" + @"\result_" + num + format;

                    if (!File.Exists(storagePath))
                        break;
                    num++;
                }

                //Image Stitching

                Stitcher.Mode mode = Stitcher.Mode.Scans;
                float threshhold = 0.0001f;
                Mat result = new Mat();

                using (Stitcher stitcher = new Stitcher(mode))                                                  //Creates a Stitcher configured with one of the stitching modes
                using (AKAZE detector = new AKAZE(AKAZE.DescriptorType.Mldb, 0, 3, threshhold))                 //Creates a Feature Detector with a user-chosen threshhold
                using (VectorOfMat input = new VectorOfMat())
                {
                    stitcher.SetFeaturesFinder(detector);

                    Mat[] images = new Mat[filePaths.Length];

                    for (int i = 0; i < filePaths.Length; i++)
                    {
                        images[i] = CvInvoke.Imread(filePaths[i], ImreadModes.AnyColor);
                    }

                    input.Push(images);

                    for (int i = 0; i < filePaths.Length; i++)
                    {
                        images[i].Dispose();
                    }

                    stitcher.Stitch(input, result);

                }

                using (Image<Bgr, Byte> resImage = result.ToImage<Bgr, Byte>())
                {
                    resImage.Save(storagePath);
                }

                result.Dispose();

                Console.ReadKey();              //I included this, so the program wont stop and you can see, how the mempry consumption graph is not falling
            }

        }
    }

[dotnetfiddle.net/x6Iy1Q](url) shows that the managed memory usage is less than 1MB - virtually all of the memory usage is unmanaged.  I am stitching images of very high resolution so i would be really thankful if this issue could be fixed.

**Comment by emgucv** [comment]
Tested with the following code using Emgu CV 4.1 release:

```
            float threshold = 0.0001f;
            using (Stitcher stitcher = new Stitcher(Stitcher.Mode.Scans))
            using (AKAZE detector = new AKAZE(AKAZE.DescriptorType.Mldb, 0, 3, threshold))
            {
                stitcher.SetFeaturesFinder(detector);

                for (int i = 0; i < 100; i++)
                {
                    Mat[] images = new Mat[4];

                    images[0] = CvInvoke.Imread("stitch1.jpg");
                    images[1] = CvInvoke.Imread("stitch2.jpg");
                    images[2] = CvInvoke.Imread("stitch3.jpg");
                    images[3] = CvInvoke.Imread("stitch4.jpg");

                    using (Mat result = new Mat())
                    using (VectorOfMat input = new VectorOfMat())
                    {
                        input.Push(images);
                        stitcher.Stitch(input, result);
                    }

                    foreach (Mat m in images)
                        m.Dispose();
                }
            }
```
Basically stitching 100 times using the 4 images from this folder:
https://github.com/emgucv/emgucv/tree/master/Emgu.CV.Test 

Memory graph looks quite normal after ~18 minutes and 40 stitching operations:

![image](https://user-images.githubusercontent.com/2035816/60934589-d1c94980-a294-11e9-8bb5-8199b5a2eec7.png)

My images are 640x480 and I didn't observer any significant memory leak.

**Comment by emgucv** [comment]
Sorry, doesn't mean to close the ticket, pressed wrong button.

**Comment by frenchcs1** [comment]
I am having the same issue with the stitcher, I am rather new to using EMGU so I can imagaine I'm doing something wrong but wanted to log the issue. I am using Visual Studio 2017 and installed EMGU.CV through Nuget. Not sure if you maintain the Nuget package or not but figured I would say where I got EMGU from.

Right now I am trying to iterate through a list of Images in two separate folders and then save the stiched image into another directory. The images I use are 1920x1080 and after the stitch I am cropping off the edges. Here is some sample code of what I am doing:

```
private static void Stitch()
{
    float threshold = 0.0001f;
    DirectoryInfo dir = new DirectoryInfo(loc);
    FileInfo[] filesP1;
    FileInfo[] filesP2;
    if (dir.Exists)
    {
        filesP1 = dir.GetFiles("P1*", SearchOption.AllDirectories).OrderBy(o => o.LastWriteTime).ToArray();
        filesP2 = dir.GetFiles("P2*", SearchOption.AllDirectories).OrderBy(o => o.LastWriteTime).ToArray();
    }
    else
    {
        throw new Exception("No such folder.");
    }

    for (int i = 0; i < filesP1.Length; i++)
    {
        Mat[] sourceImages = new Mat[2];
        sourceImages[0] = CvInvoke.Imread(filesP1[i].FullName);
        sourceImages[1] = CvInvoke.Imread(filesP2[i].FullName);

        using (Stitcher stitcher = new Stitcher(Stitcher.Mode.Panorama))
        {
            using (AKAZE finder = new AKAZE(AKAZE.DescriptorType.Mldb, 0, 3, threshold))
            {
                stitcher.SetFeaturesFinder(finder);
                using (VectorOfMat vm = new VectorOfMat())
                {
                    using (Mat result = new Mat())
                    {
                        vm.Push(sourceImages);

                        Stitcher.Status stitchStatus = stitcher.Stitch(vm, result);

                        if (stitchStatus == Stitcher.Status.Ok)
                        {
                            using (FileStream stream = new FileStream(filesP1[i].DirectoryName + $@"..\..\Stitch\P1_{DateTime.Now.Ticks}_stitch.jpg", FileMode.CreateNew))
                            {
                                var rect = new Rectangle(50, 120, 3000, 800);

                                using (Bitmap saveImg = result.Bitmap.Clone(rect, result.Bitmap.PixelFormat))
                                {

                                    saveImg.Save(stream, ImageFormat.Jpeg);
                                }
                            }
                        }
                    }
                }
            }
        }
        foreach (Mat img in sourceImages)
        {
            img.Dispose();
        }
    }
}
```

Here is also an image of my memory/cpu graphs in studio:

![ScreenClip](https://user-images.githubusercontent.com/26091160/67766519-feac9280-fa24-11e9-90d1-4dbabcb4180d.png)

Once the memory usage gets up to 2 GB any image that gets saved out is entirely black and then after a couple of those the memory and cpu just kind of flatline and nothing else gets exported. I usually get between 25 and 40 images before it dies. Running it in Release doesn't appear to affect it.

**Comment by emgucv** [comment]
There are two places where images are not disposed:

> Mat[] sourceImages = new Mat[2];

needs to be disposed within the loop

> using (Bitmap saveImg = result.Bitmap.Clone(rect, result.Bitmap.PixelFormat))

result.Bitmap is not disposed, only the clone is disposed. Need to use two "using" statement such as follows:
using (Bitmap tmp = result.Bitmap)
using (Bitmap saveImg = tmp.Clone(rect, result.Bitmap.PixelFormat))

**Comment by frenchcs1** [comment]
For the Mat array I would have thought the foreach at the end of the loop would have disposed of the images correctly. Similarly I would have expected that I have the "result" in a Using that the bitmap property would also be disposed of when the result gets disposed.

I will give the additional Bitmap using a try and let you know how it goes.

As for the Mat array, could you help point me in the right direction to dispose of that in my loop?

Thanks

**Comment by emgucv** [comment]
I have re-checked the source code you have posted and "sourceImages" are disposed close to the end of the loop. There is not need to change for the "sourceImages". Only the extra Bitmap need to be disposed. 

Btw, if you are using the same parameters for the Stitcher and AKAZE, it will be more efficient to create it only once outside of the loop and re-use it for all images. Similar to the code I have posted earlier

> Tested with the following code using Emgu CV 4.1 release:
> 
> ```
>             float threshold = 0.0001f;
>             using (Stitcher stitcher = new Stitcher(Stitcher.Mode.Scans))
>             using (AKAZE detector = new AKAZE(AKAZE.DescriptorType.Mldb, 0, 3, threshold))
>             {
>                 stitcher.SetFeaturesFinder(detector);
> 
>                 for (int i = 0; i < 100; i++)
>                 {
>                     Mat[] images = new Mat[4];
> 
>                     images[0] = CvInvoke.Imread("stitch1.jpg");
>                     images[1] = CvInvoke.Imread("stitch2.jpg");
>                     images[2] = CvInvoke.Imread("stitch3.jpg");
>                     images[3] = CvInvoke.Imread("stitch4.jpg");
> 
>                     using (Mat result = new Mat())
>                     using (VectorOfMat input = new VectorOfMat())
>                     {
>                         input.Push(images);
>                         stitcher.Stitch(input, result);
>                     }
> 
>                     foreach (Mat m in images)
>                         m.Dispose();
>                 }
>             }
> ```
> 
> Basically stitching 100 times using the 4 images from this folder:
> https://github.com/emgucv/emgucv/tree/master/Emgu.CV.Test
> 
> Memory graph looks quite normal after ~18 minutes and 40 stitching operations:
> 
> ![image](https://user-images.githubusercontent.com/2035816/60934589-d1c94980-a294-11e9-8bb5-8199b5a2eec7.png)
> 
> My images are 640x480 and I didn't observer any significant memory leak.

**Comment by frenchcs1** [comment]
Ok, so here is some more information from what you have provided me. I added the additional Disposal of that Bitmap which helped slightly but not much. However, once I moved the Stitcher outside the "for" loop, I instantly no longer had any memory issues. I was able to run through 553 stitchings with just about ~700 MB of memory being used. So I still think there is something not being cleared out of memory when the Stitcher is disposed but creating it only once is probably the correct course of action anyway.

Here's a sample memory graph after moving "using stitcher" outside the for loop and running 300 stitches on my 1920x1080 images:
![ScreenClip](https://user-images.githubusercontent.com/26091160/67970184-f8fbac00-fbe0-11e9-8747-7b332938b49d.png)
